### PR TITLE
Run vapor as non-root user

### PR DESCRIPTION
**Pull request by mattpolzin**
Closes https://github.com/vapor/template/issues/23

- Add non-root user named `vapor`.
- Switch to default of port 8080 because non-root user cannot bind to port 80.

Open Question: Are there common use-cases where files outside of the directory containing the executable (`/vapor`, the home folder for the new `vapor` user) need to be written to by Vapor apps? There probably does not _need_ to be, but I would hate to cause any confusion for folks with this change.

### CODE REVIEW

**Comment by gwynne** [comment]
Holding review until conflicts from the recent major merge to master are resolved.

**Comment by mattpolzin** [comment]
@gwynne thanks! Ready for a first pass.

File: Dockerfile [hunk 1]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 FROM swift:5.2-bionic-slim
+
+RUN useradd --user-group --create-home --home-dir /app vapor
+
 WORKDIR /app
 
 # Copy build artifacts
-COPY --from=build /build/.build/release /app
-
+COPY --from=build --chown=vapor /build/.build/release .
```
Comment (gwynne) [inline Dockerfile]:
```suggestion
COPY --from=build --chown=vapor:vapor /build/.build/release .
```

File: Dockerfile [hunk 2]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 # Uncomment the next line if you need to load resources from the `Public` directory
-#COPY --from=build /build/Public /run/Public
+#COPY --from=build --chown=vapor /build/Public ./Public
```
Comment (gwynne) [inline Dockerfile]:
```suggestion
#COPY --from=build --chown=vapor:vapor /build/Public ./Public
```

File: Dockerfile [hunk 3]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
 # Run image
 # ================================
 FROM swift:5.2-bionic-slim
+
+RUN useradd --user-group --create-home --home-dir /app vapor
```
Comment (gwynne) [inline Dockerfile:30]:
```suggestion
# Create a vapor user and group with /app as its home directory
RUN useradd --user-group --create-home --home-dir /app vapor
```

File: Dockerfile [hunk 4]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 
 # Copy build artifacts
-COPY --from=build /build/.build/release /app
-
+COPY --from=build --chown=vapor /build/.build/release .
 # Uncomment the next line if you need to load resources from the `Public` directory
-#COPY --from=build /build/Public /run/Public
+#COPY --from=build --chown=vapor /build/Public ./Public
+
+USER vapor
```
Comment (gwynne) [inline Dockerfile:39]:
```suggestion
# Ensure all further commands run as the vapor user
USER vapor
```

File: Dockerfile [hunk 5]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 
 ENTRYPOINT ["./Run"]
```
Comment (gwynne) [inline Dockerfile:41]:
```suggestion
# Start the Vapor service when the image is run, default to listening on 8080 in production environment 
ENTRYPOINT ["./Run"]
```

File: Dockerfile [hunk 6]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 FROM swift:5.2-bionic-slim
+
+RUN useradd --user-group --create-home --home-dir /app vapor
+
 WORKDIR /app
 
 # Copy build artifacts
-COPY --from=build /build/.build/release /app
-
+COPY --from=build --chown=vapor /build/.build/release /app
```
Comment (gwynne) [inline Dockerfile:35]:
```suggestion
COPY --from=build --chown=vapor:vapor /build/.build/release /app
```

File: Dockerfile [hunk 7]
```diff
@@ -26,13 +26,17 @@ RUN swift build --enable-test-discovery -c release
...
 # Uncomment the next line if you need to load resources from the `Public` directory
-#COPY --from=build /build/Public /run/Public
+#COPY --from=build --chown=vapor /build/Public /app/Public
```
Comment (gwynne) [inline Dockerfile:37]:
```suggestion
#COPY --from=build --chown=vapor:vapor /build/Public /app/Public
```

**Comment by mattpolzin** [comment]
@gwynne merge away (if you've got write perms)

**Comment by gwynne** [comment]
Answering the open question: If a Vapor app needs to write somewhere that it can't without root permissions (places like `/tmp` are usually world-writeable regardless), that app _should_ have the obligation to apply appropriate `RUN chown` etc. commands to its `Dockerfile` above and beyond what's done here. We can only accommodate so many scenarios.

**Pull request merged**